pkg/policy/l4.go

```diff
@@ -1084,13 +1111,280 @@ func addL4Filter(policyCtx PolicyContext,
 		}
 	}
 
-	resMap[key] = existingFilter
+	resMap.Upsert(p.Port, uint16(p.EndPort), string(proto), existingFilter)
 	return nil
 }
 
-// L4PolicyMap is a list of L4 filters indexable by protocol/port
-// key format: "port/proto"
-type L4PolicyMap map[string]*L4Filter
+// L4PolicyMap is a list of L4 filters indexable by port/endport/protocol
+type L4PolicyMap interface {
+	Upsert(port string, endPort uint16, protocol string, l4 *L4Filter)
+	Delete(port string, endPort uint16, protocol string)
+	ExactLookup(port string, endPort uint16, protocol string) *L4Filter
+	LongestPrefixMatch(port string, protocol string) *L4Filter
+	Detach(selectorCache *SelectorCache)
+	IngressCoversContext(ctx *SearchContext) api.Decision
+	EgressCoversContext(ctx *SearchContext) api.Decision
+	ForEach(func(l4 *L4Filter) bool)
+	Equals(t *testing.T, bMap L4PolicyMap) bool
+	Diff(t *testing.T, expectedMap L4PolicyMap) string
+	Len() int
+}
+
+// NewL4PolicyMap creates an new L4PolicMap.
+func NewL4PolicyMap() L4PolicyMap {
+	return &l4PolicyMap{
+		namedPortMap: make(map[string]*L4Filter),
+		rangePortMap: bitlpm.NewUintTrie[uint32, map[portProtoKey]*L4Filter](),
+	}
+}
+
+// InitL4PolicyMap creates an new L4PolicMap, with an initial
+// set of values. The initMap argument does not support ranges.
+func InitL4PolicyMap(initMap map[string]*L4Filter) L4PolicyMap {
```

Comment:
nits: If this is only used for testing, we could probably keep it in the testing code.

Spelling:
```suggestion
// InitL4PolicyMap creates an new L4PolicyMap, with an initial
// set of values. The initMap argument does not support ranges.
func InitL4PolicyMap(initMap map[string]*L4Filter) L4PolicyMap {
```

Typically we'd also use the parlance `NewFoo()` for creating a new `Foo`, rather than `Init(...)` which sounds like it initializes an existing object.

Reply:
>nits: If this is only used for testing, we could probably keep it in the testing code.
It's used by the envoy package, unfortunately.

> Typically we'd also use the parlance NewFoo() for creating a new Foo, rather than Init(...) which sounds like it initializes an existing object.

What about `InitNew` or `NewL4PolicyMapInit`?

Reply:
I guess the common pattern would be something like the following:

`func (l4 L4PolicyMap) WithFilters(initMap map[string]*L4Filter) L4PolicyMap { ... }` 

Then we could use builder pattern elsewhere, eg:

`pm := NewL4PolicyMap().WithFilters(...)`